Implement zAuction Approval and Approve Bid using SDK

domw30 opened the pull request:
<!-- NOTE: Please use the following template - it makes the reviewer's lives much easier! -->

[Associated Notion Card](https://www.notion.so/Improve-Accept-Bids-UX-9cfb00ed0ac34d63a9dbf88f50d26488?p=7a63dd9144ae4763b25eb0984bb9e3a7)

-----

## 1. Pull request checklist

<!-- Please make sure to do the following - your PR may not be accepted if any of these aren't completed: -->
- [x] Notion card has been moved to the Code Review column
- [x] Notion card has a link to this PR
- [x] A reviewer has been assigned to the Notion card


## 2. PR type
- [x] Feature

## 3. What is the old behaviour?
Confusing UI/UX for accept bid flow.


## 4. What is the new behaviour?
Improved accept bid flow.

https://user-images.githubusercontent.com/39112648/159201413-cbf540c7-05c5-4e5d-b373-97dfec0c3cd8.mp4


PLEASE NOTE: 

- I have updated text that states two transactions are taking place. This now reads as ```'Waiting for transaction approval. You should receive a request in your wallet.'```.

- Please check refetch of owned domain table data. I am still required to refresh the page.

domw30 commented:
@padala @colbr, updated accept bid flow ready for review. I left a couple of notes in the top PR description. Thanks guys

colbr commented on src/containers/Tables/OwnedDomainsTable/BidProvider.tsx:
```diff
@@ -0,0 +1,50 @@
+// React Imports
```
This file can be removed (I think)

colbr commented on src/containers/flows/AcceptBid/AcceptBid.constants.tsx:
```diff
@@ -0,0 +1,55 @@
+import { Step } from './AcceptBid.types';
+
+export const TITLES = {
+	[Step.Details]: { PRIMARY: 'Accept Bid' },
+	[Step.CheckingZAuctionApproval]: { PRIMARY: 'zAuction Approval' },
+	[Step.Confirmation]: { PRIMARY: 'Are you sure?' },
+	[Step.Success]: { PRIMARY: 'Bid Accepted' },
```
nit: can simplify by making each value a `string`, rather than an `object`, as there's only ever one key inside each object

colbr commented on src/containers/flows/AcceptBid/AcceptBid.constants.tsx:
```diff
@@ -0,0 +1,55 @@
+	CONFIRM_BID_AMOUNT: 'Are you sure you want to accept a bid of',
+};
+
+export const getNFTConfirmDetailsText = (domainName: string) =>
+	`and transfer ownership of ${domainName} to `;
```
I think previously the domain names in these flows are bold, so you could make this function return a `ReactNode`, i.e. `=> <>and transfer ownership of <b>{domainName}</b> to</>`

domw30 replied:
tidied this up a bit more as Eth requested some changes too :)

colbr commented on src/containers/flows/AcceptBid/AcceptBid.tsx:
```diff
@@ -0,0 +1,278 @@
+
+	const [currentStep, setCurrentStep] = useState<Step>(Step.LoadingData);
+	const [error, setError] = useState<string | undefined>();
+	const [stepTitle, setStepTitle] = useState<string>(
+		TITLES[Step.Details].PRIMARY,
+	);
```
nit: I think you could just use `TITLES[currentStep]` (after making all of the values `string` rather than `object`) directly in the `<Wizard>` `header` prop - this would save needing to `setStepTitle` on step changes.

domw30 replied:
tidied this up on accept bid & transfer ownership flow

colbr commented on src/containers/flows/AcceptBid/AcceptBid.tsx:
```diff
@@ -0,0 +1,278 @@
+				highestBid={highestBid}
+				wildPriceUsd={wildPriceUsd}
+				onClose={onClose}
+				onNext={() => onDetailsAccept()}
```
nit: should be sweet to just do `onNext={onDetailsAccept}` (same as in a couple of other spots) rather than using an anonymous function

colbr commented on src/containers/flows/AcceptBid/components/Success/Success.tsx:
```diff
@@ -0,0 +1,76 @@
+//- Components Imports
```
nit: could probably make a more generic component that handles `Success.tsx` and `Details.tsx` to reduce code duplication

colbr commented on src/containers/flows/AcceptBid/components/utils.ts:
```diff
@@ -0,0 +1,40 @@
+import { CURRENCY } from 'constants/currency';
+
+// Truncated NFT domain
+export const truncatedDomain = (domainName: string) => {
+	let domainText;
+	if (('wilder.' + domainName).length > 45) {
+		domainText = `wilder...${
+			domainName.split('.')[domainName.split('.').length - 1]
+		}`;
+		return domainText;
+	} else {
+		domainText = `${domainName}`;
+		return domainText;
+	}
+};
+
+// Truncated wallet address
+export const truncatedAddress = (walletAddress: string) => {
+	const truncatedWalletAddress = `${walletAddress.substring(
+		0,
+		4,
+	)}...${walletAddress.substring(walletAddress.length - 4)}`;
+	return truncatedWalletAddress;
+};
```
These ones are defined `lib/utils` - import them from there to reduce code duplication

colbr commented on src/containers/flows/AcceptBid/hooks/useAcceptBid.ts:
```diff
@@ -0,0 +1,66 @@
+				tx = await sdk.zauction.acceptBid(bid, library.getSigner());
+			} catch (err) {
+				console.error(err);
+				throw new Error(ERRORS.SIGNATURE);
```
I believe the error here would be the `rejected` rather than `signature failed`

colbr commented on src/containers/lists/BidList/BidList.tsx:
```diff
@@ -47,83 +62,112 @@ const BidList: React.FC<BidListProps> = ({
-	// }, [blockNumber]);
-
 	const sorted = bids
 		.slice()
-		.sort((a: Bid, b: Bid) => b.date.getTime() - a.date.getTime());
+		.sort((a: Bid, b: Bid) => Number(b.timestamp) - Number(a.timestamp));
```
Could be handy to add this as a function in `lib/utils/bids.ts` - no biggie though

colbr reviewed (changes requested):
Looking awesome, good job. Noticed one incorrect error message (`signature` vs `rejected`), other than that just nits 😊

domw30 commented:
@colbr, Hey man! I've made a number of changes since the last review. If you're not too tied up during your day today please could I request a quick review 😁 Included a loom. Refetch still seems bit iffy...

[Iteration 4 walk through](https://www.loom.com/share/b325281deb554d29a68be5f7fbeb56c0)

colbr commented on src/lib/utils/domains.ts:
```diff
@@ -63,3 +63,17 @@ export const truncateAddress = (address: string) => {
 };
+
+// Truncate domain when greater than 45 characters
+export const truncatedDomain = (domainName: string) => {
```
nit: change to verb `truncateDomain` as it's a function

colbr commented on src/lib/utils/number.ts:
```diff
@@ -22,3 +26,15 @@ export const formatByDecimalPlace = (
 };
+
+// Formatted highest bid in WILD
+export const getFormattedHighestBidAmount = (highestBid?: string) =>
+	highestBid
+		? ethers.utils.formatEther(highestBid).toString() + ` ${CURRENCY.WILD}`
+		: '';
+
+// Formatted bid amount in WILD
+export const getFormattedBidAmount = (bidAmount?: string) =>
+	bidAmount
+		? ethers.utils.formatEther(bidAmount).toString() + ` ${CURRENCY.WILD}`
+		: '';
```
nit: looks like these two are the same function - could just change it to `formatBidAmount`

colbr commented on src/containers/flows/AcceptBid/AcceptBid.tsx:
```diff
@@ -0,0 +1,302 @@
+			/>
+		),
+	};
+	console.log(isTransactionComplete);
```
`console.log` to remove

colbr reviewed (changes requested):
Looking very clean, nice work. Only requesting changes because I see you're working on getting the table to refresh on accept.

I have some ideas on the above - I'm pretty sure it's because the table is refetching straight away, and the domain hasn't been fully removed from the user's owned domains list yet. I think you'll need to add a `refetch` function to `useOwnedDomains.ts` with something like the below

```js
const refetch = () => {
     setIsLoading(true)
     setTimeout(getOwnedDomains, 3000)
}
```

This will make the table show the loading spinner for 3 seconds, then attempt to get the user's owned domains. I think this delay should sort the problem, the `3000` might need some tweaking.

colbr commented:
> @colbr, thanks mate, I've made those changes. With the refetch, I also gave similar ago as I saw how `refetch` was being used in `useBidData.ts` in `CancelBid`, but same outcome. Will keep an eye on discord thanks for raising. What dya recon, shall we merge this PR and open another for refetch when we've worked out a solution? Thanks Brett

Got a reply in Discord - looks like the subgraph can take anywhere from 5 seconds to 20 minutes to update. This is probably something that is going to be difficult to handle in the dApp, so maybe the design should be altered to let the user know changes may not appear immediately. With this in mind, I'm happy to approve this PR!
